# [Tarea 02] Ejercicios Unidad 01- A

- Dax Navarrete

https://github.com/daxrpm/numerical-methods_homeworks

## CONJUNTO DE EJERCICIOS 1
Resuelva los siguientes ejercicios, tome en cuenta que debe mostrar el desarrollo completo del

### 1. Calcule los errores absoluto y relativo en las aproximaciones de 𝑝 por 𝑝∗


#### a. $p = \pi, p^* = \frac{22}{7}$

In [6]:
import math # for pi representation

p_obj: float = 22/7
p = math.pi

print(p_obj)
print(p)

3.142857142857143
3.141592653589793


In [ ]:
abs_error = abs(p - p_obj)
abs_error


0.0012644892673496777

In [11]:
rel_error = abs((p - p_obj)/p)
rel_error

0.0004024994347707008

Dado:
$ p = \pi, \quad p^* = \frac{22}{7} $

Cálculo numérico:
$ p^* = \frac{22}{7} = 3.142857142857143 $
$ p = 3.141592653589793 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |3.141592653589793 - 3.142857142857143| $

$ E_a = 0.0012644892673496777 $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{0.0012644892673496777}{3.141592653589793} $

$ E_r = 0.0004024994347707008 $



##### Resultados
$ E_a = 0.0012644892673496777 $
$ E_r = 0.0004024994347707008 $


#### b. $ p = \pi, p^* = 3.1416$ 

In [15]:
import math

p_obj = 3.1416
p = math.pi



In [16]:
abs_error = abs(p - p_obj)
abs_error

7.346410206832132e-06

In [14]:
rel_error = abs((p - p_obj) / p)
rel_error

2.3384349967961744e-06

Dado:
$ p = \pi, \quad p^* = 3.1416 $

Cálculo numérico:
$ p^* = 3.1416 $
$ p = 3.141592653589793 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |3.141592653589793 - 3.1416| $

$ E_a = 7.346410206832132e^{-06} $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{7.346410206832132e^{-06}}{3.141592653589793} $

$ E_r = 2.3384349967961744e^{-06} $

##### Resultados
$ E_a = 7.346410206832132e^{-06} $
$ E_r = 2.3384349967961744e^{-06} $


#### c. $ p = e, p^* = 2.718 $

In [20]:
import math

p_obj = 2.718
p = math.e


In [26]:
abs_error = abs(p - p_obj)
abs_error


0.00021356237309522186

In [21]:
rel_error = abs((p - p_obj) / p)
rel_error

0.00010367889601972718

Dado: $ p = e,\quad p^* = 2.718 $

Cálculo numérico:
$ p^* = 2.718 $
$ p = 2.718281828459045 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |2.718281828459045 - 2.718| $

$ E_a = 0.00021356237309522186 $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{0.00021356237309522186}{2.718281828459045} $

$ E_r = 0.00010367889601972718 $

##### Resultados
$ E_a = 0.00021356237309522186 $
$ E_r = 0.00010367889601972718 $


#### d. $ p = \sqrt{2}, p^* = 1.414 $

In [25]:
import math

p_obj = 1.414
p = math.sqrt(2)


In [22]:
abs_error = abs(p - p_obj)
abs_error

0.0002818284590451192

In [24]:
rel_error = abs((p - p_obj) / p)
rel_error

0.00010367889601972718

Dado:
$ p = \sqrt{2}, \quad p^* = 1.414 $

Cálculo numérico:
$ p^* = 1.414 $
$ p = 1.4142135623730951 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |1.4142135623730951 - 1.414| $

$ E_a = 0.0002818284590451192 $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{0.0002818284590451192}{1.4142135623730951} $

$ E_r = 0.00010367889601972718 $

##### Resultados
$ E_a = 0.0002818284590451192 $
$ E_r = 0.00010367889601972718 $


### 2. Calcule los errores absoluto y relativo en las aproximaciones de 𝑝 por 𝑝∗


#### a

In [33]:
p = math.e**10
p_obj = 22000


In [34]:

abs_error = abs(p - p_obj)
abs_error

26.465794806703343

In [35]:
rel_error = abs_error / abs(p)
rel_error


0.0012015452253326688

Dado: $ p = e^{10},\quad p^* = 22000 $

Cálculo numérico:
$ p = e^{10} \approx 22026.465794806703 $
$ p^* = 22000 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |22026.465794806703 - 22000| $

$ E_a = 26.465794806703343 $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{26.465794806703343}{22026.465794806703} $

$ E_r = 0.0012015452253326888 $

##### Resultados
$ E_a = 26.465794806703343 $
$ E_r = 0.0012015452253326888 $


#### b.

In [39]:
import math

p = 10**math.pi
p_obj = 1400



In [40]:
abs_error = abs(p - p_obj)
abs_error

14.544268632989315

In [41]:
rel_error = abs_error / abs(p)
rel_error

0.010497822704619136

Dado: $ p = 10^{\pi},\quad p^* = 1400 $

Cálculo numérico:
$ p = 10^{\pi} \approx 1414.545408421 $
$ p^* = 1400 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |1414.545408421 - 1400| $

$ E_a = 14.544268632989315 $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{14.544268632989315}{1414.545408421} $

$ E_r = 0.010497822704619136 $

##### Resultados
$ E_a = 14.544268632989315 $
$ E_r = 0.010497822704619136 $


#### C

In [43]:
p = math.factorial(8)
p_obj = 39900

In [44]:
bs_error = abs(p - p_obj)
abs_error


14.544268632989315

In [45]:
rel_error = abs_error / abs(p)
rel_error

0.0003607209482388223

Dado: $ p = 8!,\quad p^* = 39900 $

Cálculo numérico:
$ p = 8! = 40320 $
$ p^* = 39900 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |40320 - 39900| $

$ E_a = 14.544268632989315 $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{14.544268632989315}{40320} $

$ E_r = 0.0003607209482388223 $

##### Resultados
$ E_a = 14.544268632989315 $
$ E_r = 0.0003607209482388223 $


#### d.

In [46]:
p = math.factorial(9)
p_obj = math.sqrt(18 * math.pi * ((9/math.e)**9))

In [47]:
abs_error = abs(p - p_obj)
abs_error

361235.7141751769

In [48]:
rel_error = abs_error / abs(p)
rel_error

0.9954687890629875

Dado: $ p = 9!,\quad p^* = \sqrt{18\pi \left(\frac{9}{e}\right)^9} $

Cálculo numérico:
$ p = 9! = 362880 $
$ p^* = \sqrt{18\pi(9/e)^9} \approx 361235.7141751769 $

**Error absoluto**

$ E_a = |p - p^*| $

$ E_a = |362880 - 361235.7141751769| $

$ E_a = 1644.2858248231 $

**Error relativo**

$ E_r = \frac{|p - p^*|}{|p|} $

$ E_r = \frac{1644.2858248231}{362880} $

$ E_r = 0.004528038078984 $

##### Resultados
$ E_a = 1644.2858248231 $
$ E_r = 0.004528038078984 $


### 3. Encuentre el intervalo más largo en el que se debe encontrar 𝑝∗ para aproximarse a 𝑝 con error relativo máximo de 10−4 para cada valor de 𝑝

In [49]:

def intervalo(p, tol=1e-4):
    delta = tol * abs(p)
    return (p - delta, p + delta)

In [51]:
intervalo_pi = intervalo(math.pi)
intervalo_pi

(3.141278494324434, 3.141906812855152)

In [52]:
intervalo_e = intervalo(math.e)
intervalo_e

(2.718010000276199, 2.718553656641891)

In [53]:
intervalo_sqrt2 = intervalo(math.sqrt(2))
intervalo_sqrt2

(1.4140721410168577, 1.4143549837293325)

In [55]:
intervalo_cuberoot7 = intervalo(7**(1/3))
intervalo_cuberoot7

(1.9127398896541117, 1.9131224758906662)

### 4. Use la aritmética de redondeo de tres dígitos para realizar lo siguiente. Calcule los errores absoluto y relativo con el valor exacto determinado para por lo menos cinco dígitos.

In [57]:
def sig_round(x, n=3):
    if x == 0 or math.isnan(x) or math.isinf(x):
        return x
    sign = -1 if x < 0 else 1
    x = abs(x)
    k = int(math.floor(math.log10(x)))
    factor = 10**(n - 1 - k)
    return sign * round(x * factor) / factor

#### a) $\displaystyle \frac{\frac{13}{14} - \frac{5}{7}}{2e^{-5.4}}$

$ \frac{13}{14} = 0.928571... \Rightarrow 0.929 $

$ \frac{5}{7} = 0.714285... \Rightarrow 0.714 $

$ \frac{0.929}{0.714} = 1.30112... \Rightarrow 1.30 $

$ e^{-5.4} = 0.00451658... \Rightarrow 0.00452 $

$ 2 \cdot 0.00452 = 0.00904 $

$ \frac{1.30}{0.00904} = 143.8... \Rightarrow \boxed{144} $

**Valor exacto:**  
$ p = 143.8930023 $

**Error absoluto:**  
$ E_a = |144 - 143.8930023| = 0.1069977 \Rightarrow \boxed{0.107} $

**Error relativo:**  
$ E_r = \frac{0.107}{143.8930023} = 7.43 \times 10^{-4} $


#### b) $\displaystyle -10\pi + 6e - \frac{3}{61}$

$ \pi = 3.14159... \Rightarrow 3.14 $
$ e = 2.71828... \Rightarrow 2.72 $

$ 10 \cdot 3.14 = 31.4 $
$ -31.4 $

$ 6 \cdot 2.72 = 16.32 \Rightarrow 16.3 $

$ -31.4 + 16.3 = -15.1 $

$ \frac{3}{61} = 0.049180... \Rightarrow 0.0492 $

$ -15.1 - 0.0492 = -15.1492 \Rightarrow \boxed{-15.1} $

**Valor exacto:**  
$ p = -15.1554159... $

**Error absoluto:**  
$ E_a = |-15.1 - (-15.1554159)| = 0.0554 $

**Error relativo:**  
$ E_r = \frac{0.0554}{15.1554159} = 0.00366 $


#### c) $\displaystyle \left(\frac{2}{9}\right)\left(\frac{9}{11}\right)$

$ \frac{2}{9} = 0.222222... \Rightarrow 0.222 $

$ \frac{9}{11} = 0.818181... \Rightarrow 0.818 $

$ 0.222 \cdot 0.818 = 0.181596... \Rightarrow \boxed{0.182} $

**Valor exacto:**  
$ p = \frac{2}{11} = 0.181818... $

**Error absoluto:**  
$ E_a = |0.182 - 0.181818| = 0.000182 $

**Error relativo:**  
$ E_r = \frac{0.000182}{0.181818} = 0.00100 $


#### d) $\displaystyle \frac{\sqrt{13}+\sqrt{11}}{\sqrt{13}-\sqrt{11}}$

$ \sqrt{13} = 3.60555... \Rightarrow 3.61 $
$ \sqrt{11} = 3.31662... \Rightarrow 3.32 $

$ 3.61 + 3.32 = 6.93 $
$ 3.61 - 3.32 = 0.29 $

$ \frac{6.93}{0.29} = 23.89655... \Rightarrow \boxed{23.9} $

**Valor exacto:**  
$ p = 23.9582607... $

**Error absoluto:**  
$ E_a = |23.9 - 23.9582607| = 0.0583 $

**Error relativo:**  
$ E_r = \frac{0.0583}{23.9582607} = 0.00243 $


### 5

In [58]:
import math

def arctan_poly(x):
    return x - (x**3)/3 + (x**5)/5

approx_a = 4 * (arctan_poly(1/2) + arctan_poly(1/3))
exact_a = math.pi
abs_err_a = abs(approx_a - exact_a)
rel_err_a = abs_err_a / abs(exact_a)

approx_b = 16 * arctan_poly(1/5) - 4 * arctan_poly(1/239)
exact_b = math.pi
abs_err_b = abs(approx_b - exact_b)
rel_err_b = abs_err_b / abs(exact_b)

approx_a, exact_a, abs_err_a, rel_err_a, approx_b, exact_b, abs_err_b, rel_err_b


(3.1455761316872426,
 3.141592653589793,
 0.003983478097449478,
 0.0012679804598147663,
 3.1416210293250346,
 3.141592653589793,
 2.837573524150372e-05,
 9.032277055104427e-06)



#### a) $4[\arctan(1/2) + \arctan(1/3)]$

Aproximación:
$ p^* = 3.1455761316872426 $

Valor real:
$ p = \pi = 3.141592653589793 $

Error absoluto:
$ E_a = |3.1455761316872426 - 3.141592653589793| = 0.00398347809744978 $

Error relativo:
$ E_r = \frac{0.00398347809744978}{3.141592653589793} = 0.0012679804950147663 $

---

#### b) $16\arctan(1/5) - 4\arctan(1/239)$

Aproximación:
$ p^* = 3.141629203250346 $

Valor real:
$ p = \pi = 3.141592653589793 $

Error absoluto:
$ E_a = |3.141629203250346 - 3.141592653589793| = 3.25556605527036 \times 10^{-5} $

Error relativo:
$ E_r = \frac{3.25556605527036 \times 10^{-5}}{3.141592653589793} = 9.032277055104427 \times 10^{-6} $


### 6

In [59]:

def approx_e(n):
    return sum(1/math.factorial(k) for k in range(n+1))

exact = math.e


approx_a = approx_e(5)
abs_err_a = abs(approx_a - exact)
rel_err_a = abs_err_a / abs(exact)

approx_b = approx_e(10)
abs_err_b = abs(approx_b - exact)
rel_err_b = abs_err_b / abs(exact)

approx_a, exact, abs_err_a, rel_err_a, approx_b, exact, abs_err_b, rel_err_b


(2.716666666666667,
 2.718281828459045,
 0.0016151617923783057,
 0.0005941848175815963,
 2.7182818011463845,
 2.718281828459045,
 2.7312660577649694e-08,
 1.0047766310211053e-08)



#### a) $ \displaystyle \sum_{n=0}^{5} \frac{1}{n!} $

Aproximación:
$ p^* = 2.716666666666667 $

Valor real:
$ p = e = 2.718281828459045 $

Error absoluto:
$ E_a = |2.716666666666667 - 2.718281828459045| = 0.001615161792378057 $

Error relativo:
$ E_r = \frac{0.001615161792378057}{2.718281828459045} = 0.0005941481758158963 $

---

#### b) $ \displaystyle \sum_{n=0}^{10} \frac{1}{n!} $

Aproximación:
$ p^* = 2.7182818011463845 $

Valor real:
$ p = e = 2.718281828459045 $

Error absoluto:
$ E_a = |2.7182818011463845 - 2.718281828459045| = 2.731266307549694e^{-08} $

Error relativo:
$ E_r = \frac{2.731266307549694e^{-08}}{2.718281828459045} = 1.004776631201503e^{-08} $


### 7

In [60]:


def sig_round(x,n=3):
    if x==0: return 0.0
    sign = -1 if x<0 else 1
    x = abs(x)
    k = int(math.floor(math.log10(x)))
    factor = 10**(n-1-k)
    return sign * round(x*factor) / factor

def R(x,n=3): 
    return sig_round(x,n)

x0, y0 = 1.31, 3.24
x1, y1 = 1.93, 5.76
n = 3


t1 = R(R(x0)*R(y1))
t2 = R(R(x1)*R(y0))
num1 = R(t1 - t2)
den = R(R(y1) - R(y0))
x_formula1 = R(num1 / den)


diff = R(R(x1) - R(x0))
num2 = R(diff * R(y0))
x_formula2 = R(R(x0) - R(num2 / den))

x_exact = (x0*y1 - x1*y0)/(y1-y0)

x_formula1, x_formula2, x_exact


(0.516, 0.512, 0.5128571428571429)

### Ejercicio 7

Se tienen dos puntos:
$ (x_0, y_0) = (1.31, 3.24), \quad (x_1, y_1) = (1.93, 5.76) $

Con aritmética de redondeo a tres cifras.

---

#### Fórmula 1
$$ x = \frac{x_0y_1 - x_1y_0}{y_1 - y_0} $$

$ x_0 y_1 = 1.31 \cdot 5.76 = 7.5456 \Rightarrow 7.55 $  
$ x_1 y_0 = 1.93 \cdot 3.24 = 6.2532 \Rightarrow 6.25 $  
$ 7.55 - 6.25 = 1.30 $  
$ y_1 - y_0 = 5.76 - 3.24 = 2.52 $  
$ x = 1.30 / 2.52 = 0.51587... \Rightarrow \boxed{0.516} $

---

#### Fórmula 2
$$ x = x_0 - \frac{(x_1 - x_0)y_0}{y_1 - y_0} $$

$ x_1 - x_0 = 1.93 - 1.31 = 0.62 $  
$ 0.62 \cdot 3.24 = 2.0088 \Rightarrow 2.01 $  
$ 2.01 / 2.52 = 0.7976... \Rightarrow 0.798 $  
$ x = 1.31 - 0.798 = 0.512 \Rightarrow \boxed{0.512} $

---

#### Valor exacto (sin redondeo)
$$ x_{\text{exacto}} = \frac{x_0y_1 - x_1y_0}{y_1-y_0} = 0.5128571428571429 $$

---

### Comparación

| Método | Resultado | Error Absoluto |
|-------|-----------|----------------|
| Fórmula 1 | 0.516 | $0.516 - 0.512857 = 0.00314$ |
| Fórmula 2 | 0.512 | $0.512 - 0.512857 = 0.000857$ |

---

### Conclusión
La segunda fórmula es mejor porque evita hacer tantas operaciones que puedan aumentar la propagacion de errores



> **Nota** declaro el uso de IA(modelo GPT5) para explicacion de algunos ejercicios y para ayuda con markdown e implementaciones de codigo
